In [1]:
import numpy as np
import matplotlib.pyplot as plt

# 1.10
# Slot machine reward distribution
rewards = [0, 5, 10, 20]  # Dollar amounts
probabilities = [0.35, 0.3, 0.25, 0.1]  # Probabilities

#Second slot machine with new reward distribution
rewards2 = [0, 5, 10, 20]  # Dollar
probabilities2 = [0.3, 0.35, 0.3, 0.05]  # New Probabilities

# Calculate the distribution of the sum
sum_distribution = {}
for r_reward, r_prob in zip(rewards, probabilities):
    for s_reward, s_prob in zip(rewards2, probabilities2):
        total = r_reward + s_reward
        prob = r_prob * s_prob
        if total not in sum_distribution:
            sum_distribution[total] = 0
        sum_distribution[total] += prob

# Sort by total reward
sorted_sum = sorted(sum_distribution.items())
sum_totals = [x[0] for x in sorted_sum]
sum_probs = [x[1] for x in sorted_sum]

print("Distribution of the Sum of Rewards P(R+S):")
print(f"{'Sum of Rewards':<18} {'Probability':<15}")
print("-" * 35)
for total, prob in sorted_sum:
    print(f"${total:<17} {prob:.4f} ({prob:.2%})")



Distribution of the Sum of Rewards P(R+S):
Sum of Rewards     Probability    
-----------------------------------
$0                 0.1050 (10.50%)
$5                 0.2125 (21.25%)
$10                0.2850 (28.50%)
$15                0.1775 (17.75%)
$20                0.1225 (12.25%)
$25                0.0500 (5.00%)
$30                0.0425 (4.25%)
$40                0.0050 (0.50%)


Exercise 1.14 (*)

Since both slot machines are independent, the probability of the sum of both machines being more than 20 is:
P(R+S > 20 | R = 10) = P(S > 10) = P(S = 20) = 0.05

Exercise 4.2

States: we can map the surfaces that needs to be cleaned as an x,y grid where each cell represents a part of the room. The state would then be the current position of the robot on this grid and which cells have been cleaned.

Actions: The actions would be the possible movements the robot can make (up, down, left, right) and an action to vacuum the current cell.

Rewards: The robot would recieve a positive reward for vacuuming a cell that is dirty and a small negative reward for each movement to encourage efficiency. As it is a house cleaning robot, we can also impose a negative reward for crashing into obstacles or walls.

Exercise 4.6

The reason for the competing robot crashing more, but performing better in other ways, can be that the competing firm has prioritized performance (time, efficient sweeping) over safety (avoiding crashes). 
It could for example be that the competing robot drives faster, which increases the chance of crashing, but also allows it to finish the task quicker.
For sure the competing robot has a less negative reward for crashing, which leads it to adapt policies that crash more often.

Exercise 5

We find the choice to use RL in this scenario inappropriate as it would face some ethical and practical challenges.
Firstly, RL typically requires a large number of interactions with the environment to learn an effective policy. This is ethically problematic as it would risk the first applicants to not be considered fairly, as the system would be learning and making mistakes during this period.

Secondly, the recruitment process often involves subjective judgments and qualitative factors that are difficult to represent and incorporate into a reward function. E.g. it would be difficult to design well-structured reward functions that capture all relevant aspects of a candidate's potential. One could argue that you can create some kind of efficiency measure, but the data collection for this would likely take a long time and be very noisy.



Exercise 6. MABs

Giving three different vaccines to the same population and gradually estimating which vaccine is the most efficient based on a relevant reward measure. To get these estimates you would need to use exploration vs exploitation strategies to make sure you get good estimates of each vaccine's efficiency. A good strategy would ensure that you gradually would shift towards giving the most efficient vaccine to the majority of the population and thus maximize your goal.

This MAB approach can be more sensible than a RCT (randomized control trial) as it is adaptive and will gradually give more people the best vaccine. This can be crucial in for example a pandemic situation where you want to maximize the efficiency of the vaccine rollout as quickly as possible and not wait for the results of a full RCT before changing strategy.
